File: ad_library/src/main/java/com/adgrowth/adserver/exceptions/SDKInitException.java
package com.adgrowth.adserver.exceptions;

public class SDKInitException extends Exception {

    public static final String UNAUTHORIZED_CLIENT_KEY = "unauthorized_client_key";
    public static final String ALREADY_INITIALIZED = "already_initialized";
    public static final String INTERNAL_ERROR = "internal_error";
    public static final String NETWORK_ERROR = "network_error";
    public static final String UNKNOWN_ERROR = "unknown_error";
    private String mCode;
    private String message;

    public SDKInitException(AdRequestException e) {
        handleError(e.getCode());
    }

    public SDKInitException(String errorCode) {
        mCode = errorCode;
        setMessage(mCode);
    }

    private void handleError(String errorCode) {
        switch (errorCode) {
            case AdRequestException.NETWORK_ERROR:
                mCode = NETWORK_ERROR;
                break;
            case AdRequestException.INTERNAL_ERROR:
                mCode = INTERNAL_ERROR;
                break;
            case AdRequestException.UNKNOWN_ERROR:
            default:
                mCode = UNKNOWN_ERROR;
                break;
        }

        setMessage(mCode);
    }

    private void setMessage(String code) {
        switch (code) {

            case ALREADY_INITIALIZED:
                message = "You already initialized this with this key, just use it.";
                break;
            case UNAUTHORIZED_CLIENT_KEY:
                message = "The provided client key is not valid.";
                break;
            case UNKNOWN_ERROR:
            case INTERNAL_ERROR:
                message = "Sorry, this was an unexpected error, please try again.";
                break;
            default:
        }
    }

    public String getCode() {
        return mCode;
    }

    public String getMessage() {
        return message;
    }
}

File: ad_library/src/main/kotlin/com/adgrowth/adserver/exceptions/SDKInitException.kt
package com.adgrowth.adserver.exceptions

class SDKInitException : Exception {
    var code: String? = null
        private set
    override var message: String? = null
        private set(code) {
            when (code) {
                INVALID_CLIENT_KEY -> field = "Your app key does not match any app in our database."
                APP_NOT_FOUND -> field = "Your app key doesn't match with any app on our database."
                ALREADY_INITIALIZED -> field ="You already initialized this with this key, just use it."
                UNAUTHORIZED_CLIENT_KEY -> field = "The provided client key is not valid."
                UNKNOWN_ERROR, INTERNAL_ERROR -> field = "Sorry, this was an unexpected error, please try again."
                NETWORK_ERROR -> field = "It is possible that your internet connection is unstable, please try again."
                else -> {}
            }
        }

    constructor(e: AdRequestException) {
        code = e.code
        message = e.code
    }

    constructor(errorCode: String?) {
        this.code = errorCode
        message = this.code
    }


    companion object {
        const val UNAUTHORIZED_CLIENT_KEY = "unauthorized_client_key"
        const val ALREADY_INITIALIZED = "already_initialized"
        const val APP_NOT_FOUND = "app_not_found"
        const val INVALID_CLIENT_KEY = "invalid_client_key"
        const val INTERNAL_ERROR = "internal_error"
        const val NETWORK_ERROR = "network_error"
        const val UNKNOWN_ERROR = "unknown_error"
    }
}
